### ISSUE TITLE
missing instructions in Node Map generation

### ISSUE THREAD

**Issue by pchampin** [w3c/json-ld-api#549]
I believe that an instruction is missing from step 6.10 of [Node Map Generation](https://www.w3.org/TR/json-ld11-api/#node-map-generation), namely:

- if <var>node map</var> does not contain an <var>id<var> entry, create one and initialize its value to an empty map

before recursively invoking the algorithm.
Indeed, in step 2 of the invoked algorithm, that entry is assumed to already exist:

> Reference the [map](https://infra.spec.whatwg.org/#ordered-map) which is the value of the [active graph](https://www.w3.org/TR/json-ld11-api/#dfn-active-graph) [entry](https://infra.spec.whatwg.org/#map-entry) of node map")

**Comment by gkellogg** [comment]
Sorry, a bit confused. Step 6.10 is the following:

> 6.10) If _element+ has an `@graph` [entry](https://infra.spec.whatwg.org/#map-entry), recursively invoke this algorithm passing the value of the `@graph` [entry](https://infra.spec.whatwg.org/#map-entry) for _element_, _node map_, and _id_ for _active graph_ before removing the `@graph` [entry](https://infra.spec.whatwg.org/#map-entry) from _element_.

In fact, when searching the document (both TR and ED), I can't find a sentence "if node map does not contain an id entry, create one and initialize its value to an empty map".

The algorithm takes _node map_ as an input argument, so it definitely must already exist. The places which recursively invoke the algorithm pass the existing _node map_, so I don't see the specific issue.

**Comment by pchampin** [comment]
The sentence I wrote ("if node map does not contain an id entry...") is what I suggest to add to step 6.10 . So indeed, it is currently not in the spec :-)

Of course <var>node map</var> always exists. What does not always exist, when entering 6.10,  is an entry in <var>node map</var> whose key is <var>id</var>. When the algorithm is recursively invoked, passing <var>node map</var> and passing <var>id</var> for <var>active graph</var>, this entry is assumed to exist at step 2.

By adding the suggested sentence in 6.10 before the recursive invocation, we ensure that this assumption in the invoked step 2 is satisfied.

**Comment by niklasl** [comment]
I agree that what @pchampin suggests seems to be needed. (I've made a [similar note](https://github.com/niklasl/trld/blob/0.2.0/trld/jsonld/flattening.py#L95) in my implementation.) 

I've later on also noticed that this [node map generation algorithm](https://w3c.github.io/json-ld-api/#algorithm-10) _mutates_ the input. Several steps (6.7, 6.9, 6.10, 6.11) says to _remove_ an entry from the element.  Therefore, I think step 2, where element is a map, also needs to make a (shallow) copy of this element, which is then subsequently mutated. (In 6.12, the remaining keys of the element are used, so there are other ways of fixing this; making a copy is just the [easiest change](https://github.com/niklasl/trld/commit/b939085ec3ebe4bc0701592d369bcc932918082b) to the current design.)
